File: frontend/src/types/auditLog.ts
export interface ActorInfo {
  id: string;
  email: string;
  first_name?: string | null;
  last_name?: string | null;
}

export interface AuditLog {
  id: string;
  organization_id?: string | null;
  actor_id?: string | null;
  actor?: ActorInfo | null;
  action: string;
  resource_type: string;
  resource_id?: string | null;
  details?: Record<string, any> | null;
  ip_address?: string | null;
  user_agent?: string | null;
  created_at: string;
}

export interface AuditLogListParams {
  page?: number;
  limit?: number;
  search?: string;
  resource_type?: string;
  action?: string;
  from_date?: string;
  to_date?: string;
  sort_by?: string;
  sort_order?: "asc" | "desc";
}

export interface AuditLogListResponse {
  items: AuditLog[];
  total: number;
  page: number;
  limit: number;
  total_pages: number;
}

File: backend/app/schemas/audit_log.py
import uuid
from datetime import datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field


class ActorInfo(BaseModel):
    """Actor information snapshot."""

    id: uuid.UUID
    email: str
    first_name: Optional[str] = None
    last_name: Optional[str] = None

    model_config = {"from_attributes": True}


class AuditLogResponse(BaseModel):
    """Single audit log entry response DTO."""

    id: uuid.UUID
    organization_id: Optional[uuid.UUID] = None
    actor_id: Optional[uuid.UUID] = None
    actor: Optional[ActorInfo] = None
    action: str = Field(..., description="Action name (e.g. user.login, project.created, api_key.regenerated)")
    resource_type: str = Field(..., description="Target resource type (e.g. user, project, api_key, settings)")
    resource_id: Optional[str] = None
    details: Optional[Dict[str, Any]] = None
    ip_address: Optional[str] = None
    user_agent: Optional[str] = None
    created_at: datetime

    model_config = {"from_attributes": True}


class AuditLogListResponse(BaseModel):
    """Paginated list of audit log entries."""

    items: List[AuditLogResponse]
    total: int
    page: int
    limit: int
    total_pages: int
